### DISCUSSION TITLE
ARM Sessions for Drives (requesting community feedback)

### DISCUSSION THREAD

**Discussion by microtechno9000** [automatic-ripping-machine/automatic-ripping-machine#815] [Ideas]
All

There has been a few issues raised recently around the flexibility of ARM and an ability to create or edit more that the standard DVD, Bluray and Music settings. I agree, and some more flexibility and better states for managing things ripping 'sessions' would be good. I personally have run into issues with Tv Show naming, ripping home movies stored on DVD's with bad titles or wanting to rip MP3's and flac at the same time from two different drives.

To solve this, I have a proposed solution, utilising ripping 'sessions' within ARM based off the existing data types DVD, Bluray, Music, Data and ISO; but expanded out for users to then edit. This will minimise the changes required to the ripper code to achieve the desired outcome.

ARM Default data types - not user editable, based on the current ARM architecture
User Sessions - User editable configuration for sessions, an example set below. User can create, read, update and delete any session

Before starting down the path of implementing this, rather large change, I would like feedback from the ARM users on what they would like in such an update. If you could respond, with any pro's/con's/I can help statements, I will compile them below

**Behaviour**
Edited to add behaviour. These items will be added into the [ARM project ](https://github.com/orgs/automatic-ripping-machine/projects/6) as tasks for the developers (or anyone) to implement.
The idea with implementing each of these will be a gradual progress towards the ARM sessions.
- ARM default data types will not change, these will be defined within the ARM code base as specific methods of ripping media **[new]**
- ARM User Sessions will be defined within the ARM code base but within the initial database load and configurable by the user to change specific fields as required **[new]**
- Ripping a movie - will rip a movie as the current ARM does **[update]**
- Ripping a TV Show will rip similar to a movie; however, using the session a User will specify the TV series to be ripped and which season is being ripped. Noting that a lot of TV series have strange disk names. Using data from movie database, ARM will use the TV series length to find the number of episodes on the disk and commence ripping. The user will be able to specify which episode the session is starting at, and then the format of the episode naming **[new]**
- Add in TV Series episode name convention. For example SERIES E1E02, Series_s01_e02, or others as defined by the user **[new]**
- Ripping music will be similar to current, but allow for multiple abcde configurations that can rip with different settings, i.e. MP3, FLAC **[new]**
- Ripping media like a home movie, or some other content that needs transcoding but with no media lookup **[new]**
- Ripping/copying data with no media look up or transcoding **[update]**
- Ripping/copying data as an ISO with no media lookup **[update]**

Overview of what some of the settings and sessions could look like.
**ARM Default Data Types**
- DVD: ARM will rip new media as a movie
- Bluray: ARM will rip new media as a blueray movie
- Music: ARM will rip new media as a music
- Data: ARM will rip new media as a data disk (copy contents)
- ISO: ARM will rip new media as an ISO image of the disk

**User Sessions**
- Movie - DVD: ARM session to rip DVD movies - Type: DVD
- Movie - Blueray: ARM sessionto rip Bluray movies - Type: Bluray
- TV - DVD: ARM session to rip DVD TV Series - Type: DVD
- TV - Blueray: ARM session to rip Bluray TV Series - Type: Bluray
- Music - MP3 : ARM session to rip music as mp3 - Type: Music
- Music - flac: ARM session to rip music as flac - Type: Music
- Data: ARM session to rip data disk - Type: Data
- HomeMovie - DVD: ARM session to rip dvd contents but with no title lookup - Type: DVD
- HomeMovie - Bluray: ARM session to rip dvd contents but with no title lookup - Type: Bluray
- ISO: ARM session to create ISO - Type: iso

**Example UI changes**
Sessions listed against the ARM system drives - from [microtechno9000 fork](https://github.com/microtechno9000/automatic-ripping-machine/tree/sessions)
![arm_sessions_drives](https://user-images.githubusercontent.com/62650032/230104782-76f405fa-907e-4848-ab3b-9889609200ce.png)

User editable session / configuration
![arm_sessions_edit](https://user-images.githubusercontent.com/62650032/230104790-a2441e99-df04-4601-8a21-60989e22a55d.png)

Help file, explaining how to use sessions and what they do
![arm_sessions_help](https://user-images.githubusercontent.com/62650032/230104798-0266fa2a-d305-40f7-95bb-5abc6d5d9eb1.png)

**Comment by Guyverix** [comment]
That would be helpful, but if it is on a per session basis, I am guessing that there is no persistence.  What about something akin to a button which would load a different yaml file on demand which would change the settings within arm?  I am guessing that would negate the need to have to worry about db changes if it is just reloading values over what has been defined.

Something like: /opt/arm/configs/custom/<file1|file2|file3>.yml  and the dropdown would allow a choice of whatever is found in that directory.  At that point, overwrite whatever is in the "ripper settings" that was already loaded with the values from the file.

**Reply by microtechno9000** [reply]
Sorry should clarify, the session would be persistent till the user turns it off.
the default behaviour would be the same as ARM now, then if a user selected a session against a drive it would remain till disabled or changed

the suggestion you made with different config files is essentially the same thing U am proposing but using a database rather than config files. The database isn't hard to update with the UI. 

both changes would require some changes to the ripping code, although your implementation would be easier/smaller.

**Comment by Guyverix** [comment]
Would end users be able to make their own custom sessions?  That way when we DO goof things up, we have a known good session to refer back to :).

**Reply by microtechno9000** [reply]
that would be the idea yes

**Comment by 1337-server** [comment]
This looks great!

**Comment by Guyverix** [comment]
I agree!  This looks really slick!

**Comment by shitwolfymakes** [comment]
I like this idea a lot. It mirrors the sessions of IDEs like PyCharm and Qt Creator

**Comment by drlauridsen** [comment]
Great idea.
I would especially like the media to iso  option..

**Comment by microtechno9000** [comment]
Slowly been making progress on sessions, current thinking will follow this model. Welcome any thoughts on the process or if anything is missing.
![Arm_session_model](https://github.com/automatic-ripping-machine/automatic-ripping-machine/assets/62650032/61dc9fa2-9715-4805-a71c-a16913c1679b)

**Comment by anthonylavado** [comment]
A comment on "TV Series episode name convention". Of course a fully customizable field for this would ultimately be the best choice that satisfies everyone. That said, I would recommend using the same defaults/parameters/variables that popular media managers (like Sonarr) have available.

**Reply by microtechno9000** [reply]
Agree, intent of all these fields is to make it user configurable. Not every one likes or uses the same naming convention.

**Comment by trleith** [comment]
Are you saying I could tell it "This session is for IMDB tt0220924" and it would automatically know "New York: A Documentary Film" has eight episodes, one per disc, grab the names and build a proper Plex TV Series folder out of it? 
```
New York - A Documentary Film (1999) {imdb-tt0220924}
     New York - A Documentary Film (1999) {imdb-tt0220924} - S1.E1 - The Country and the City (1609-1825).mkv
     New York - A Documentary Film (1999) {imdb-tt0220924} - S1.E2 - Order and Disorder (1825-1865).mkv
     ...
     New York - A Documentary Film (1999) {imdb-tt0220924} - S1.E8 - The Center of the World (1946-2003).mkv
```

This could be very useful. I have a couple TV series, one on DVD the other on Blu Ray I could test with. The Blu Ray series is six seasons of thirteen episodes, and getting the first season backed-up (with the specials/extras) was painful. I notice neither tvdb nor imdb have the extras. How can that be handled? Any idea?

**Reply by microtechno9000** [reply]
that is the plan, similarly I have TV Series that I havent ripped because ARM messes them up.

At the moment it is still in work, and aiming to get it done for ARM rev 3.0. Stay on the lookout for 2024

**Reply by DZaremba14600** [reply]
I've had the same issues with TV series.  For example, I got the first two seasons of I Love Lucy, classic episode for my parents.  There are two versions of each episode (one with the classic heart and script opening,  and one with the original cigarette sponsor's opening), plus extras.  It is such a pain trying to put them in order and have a name for the episodes I haven't even started.  I would be outstanding if the  IMDB tt code could be specified or looked up and have some of the grunt work automated in some way.  Anything would be better.

If that's that plan, that's great.

**Reply by trleith** [reply]
Where can you get information about what's on a DVD/Blu Ray/UHD? Is there something like MusicBrainz for video media?

If so, might it have proper chapter titles at least for the main features? 

This is something else that's been driving me nuts -- it reminds me of ripping CDs circa 1995.

**Comment by Guyverix** [comment]
Really looking forward to the new version!  thanks for all the work on this, it is a great tool!

**Comment by DZaremba14600** [comment]
In addition to DVD and Bluray, would there be any thoughts or objection regarding adding another data type for 4K Ultra HD?  I have found I have different settings related to 4K disks similar to the difference between DVD and Bluray.

Perhaps there are issues I'm not considering initially.  Just a thought.

Thanks.

**Reply by microtechno9000** [reply]
Is it using MKV? if so, it would support any config the user enters for MKV or Handbrake, with any number of sessions the user can create or edit

**Reply by DZaremba14600** [reply]
I'm not sure I understand - is what using MKV?  I'm using ARM as it comes.

I'd like to be able to specify different MKV and HB setting for DVD's, Bluray's, and 4k Ultra HD's.  That way I can just feed disks.  I've already had wasted sessions because I forgot to change bluray presets to 4k presets and back again as I was feeding disks.  Its MY fault for grabbing the bluray disk out of the 4k package by mistake, so I need to be more careful, but it could have settings set once and that's it.

Just my 2 cents.

But this is a great piece of work, by what I can tell is a great team.  Thanks for everything so far and not only the work but the thought that is going into this.

Much thanks!

**Comment by thatguy-jaysenodell** [comment]
Has there been any thought to moving makemkv and renaming+handbrake into separate microservices that ARM feeds jobs into? I'm thinking more of a k3s approach where 

- A queue manager owns logic "what is ripped and how does it get processed"
- makemkv is a pod ripping to a common store
- what you are referring to as sessions become pods for renaming+handbrake configurations
- rename/hb pods are fed makemkv output via service calls
- If anyfailue in process, queue manager holds rip and waits for reporcess directive.
- At last complete queue manager disposes of rip.

With this a disc could be sent to multiple services based on rules resulting in multiple output locations/formats from a single rip session. k3s resource limits would be used to manage the prioritization of processing (more resources to media server main format, fewer to portable device formats). This would also allow for easier "farms". 

This assumes we agree that using k3s to scale services over multiple machines is easy. I'm not sure many would agree.

**Reply by microtechno9000** [reply]
There has yes, not as you described, but similar. At the moment the main focus from the Devs is sorting out the ripper code, as it is a little bit of spaghetti at the moment, and also working through the Sessions idea.

Once that is done, we will be able to do things like split out portions of code or functions and create new ideas, as your describe. 

Any help with the code is always welcome, so if you have and idea, feel free to implement.

**Reply by thatguy-jaysenodell** [reply]
I may be a bit old fashion in my thoughts so forgive my use of old-timer lingo... Are you taking a refactor or a scratch approach to this? If there is a move away from a monolithic stack toward a brokered queue microservice a refactor might take longer than a scratch approach.

It's been a bit since I did much more than review code and fix random errors. Let me make some time and see what I can do. No promises.

**Reply by shitwolfymakes** [reply]
Yeah, doing a refactor to that level ain't happening until we get the UI, ripper, and data broken out into their own microservices, which is currently in development.

Additionally, something like what you propose is likely the ideal technical solution to work towards, but it's waaay overkill at the moment. We're striving for a single file that someone who's never worked with docker can take, input the filepaths into, and run with it. That is also critical to us helping keep user troubleshooting as simple as possible.

**Comment by kubycsolutions** [comment]
Proposal looks good to me.

Just to put this on the table: Samsung had a VHS/DVDR unit which could automagically copy (unprotected) tapes onto DVD among its other functions; I have one, though it's being a bit flaky. There are two formats it can record into, basically open-session (allowing its use as a DVR putting multiple episodes on one disk) or closed-session for better compatibility with other DVD players. 

Initially I've been using the former, and I've found that the resulting disks seem to play well in PC-based players. However, even if I set the the title/track-title on the DVD, ARM saves these as 

`dvds/completed/unidentified/SAMSUNG_DVD_RECORDER_VOLUME/SAMSUNG_DVD_RECORDER_VOLUME.iso`

These might (or might not) be worth addressing as part of this proposal. I can run any tests that would help. I will also try recording in the closed format and see if ARM is happier about it, if I can cure the beast's current hiccups. (Hopefully just needs dust blown out of the DVDR drive.)

_**Tangential, probably should be separate work item if it sounds reasonable:**_

I'm slightly concerned about whether disks taking a default name (device name, as here) might overwrite each other's file when ripped by ARM. Also, if unidentified stuff piles up I'm concerned about losing the order information which might tell me which disks they were out of a pile so I could rename them manually. So:

1) Should the disk ID checksum be included in the file name, so it's clear whether the user is overwriting an existing recording or making a new one with a colliding name?

2) Or, failing that, should unidentified rips include a counter or other disambiguation in their filename to ensure against unintended overwrite?

**Reply by microtechno9000** [reply]
There a few devices out there that generate specific formats, or structures that current ARM doesn't support.

As we move to the Sessions style structure, it will allow much easier modification of ARM to accommodate different data types and formats.
